lib/fog/aws/models/storage/file.rb

```diff
@@ -294,6 +308,44 @@ def multipart_save(options)
           service.complete_multipart_upload(directory.key, key, upload_id, part_tags)
         end
 
+        def multipart_copy(options, upload_part_options, target_directory_key, target_file_key)
+          # Initiate the upload
+          res = service.initiate_multipart_upload(target_directory_key, target_file_key, options)
+          upload_id = res.body["UploadId"]
+
+          # Store ETags of upload parts
+          part_tags = []
+
+          current_pos = 0
+
+          # Upload each part
+          # TODO: optionally upload chunks in parallel using threads
+          # (may cause network performance problems with many small chunks)
+          while current_pos < self.content_length do
+            start_pos = current_pos
+            end_pos = [current_pos + self.multipart_chunk_size, self.content_length - 1].min
+            range = "bytes=#{start_pos}-#{end_pos}"
+
+            upload_part_options['x-amz-copy-source-range'] = range
+            part_upload = service.upload_part_copy(target_directory_key, target_file_key, upload_id, part_tags.size + 1, upload_part_options)
+            part_tags << part_upload.body['ETag']
+            current_pos = end_pos + 1
+          end
+
+          if part_tags.empty? #it is an error to have a multipart upload with no parts
```

Comment:
I think we can get rid of this section because this should never happen.

Reply:
Yeah, not sure. We can always remove it later (though that seems unlikely to actually happen, as we are apt to forget).